### PR TITLE
Synchronous env detection via `<WithEnvironment />`

### LINKED ISSUE #59: A simpler way to detect Preact
Hi! I was doing some digging into the source code and I noticed that the current approach of detecting Preact environment creates some constraints in the other parts of the library. If we just had some simpler, ideally synchronous way of Preact detection we would have:
1. Avoided having an invisible `span` rendered by `<FpjsProvider />`. While it doesn't seem like a big deal, there could be some clients that expect the library doesn't produce any html, and could rely on some exotic CSS in their apps like `body > div`.
2. Simplified the client [initialization and loading logic](https://github.com/fingerprintjs/fingerprintjs-pro-react/blob/main/src/components/fpjs-provider.tsx#L94) and potentially avoided extra rendering at the startup. 

I don't have a definitive answer yet, however I have compiled a list of tricks that might work here:
1. A more hacky and less stable option: `ctx = createContext({}); isPreact = ctx._id || ctx.__c` (the property name is `_id` in the source code, but it becomes `__c` [after mangling](https://github.com/preactjs/preact/blob/master/mangle.json#L61)). Pros: works for Preact starting from 10.0.0 (the first release where hooks became generally available), doesn't require any additional component, completely synchronous. Cons: [it's implementation specific](https://github.com/preactjs/preact/blob/10.0.0/src/create-context.js#L13) and might break in the future.
2. Check the [number of arguments passed down to class component's `render()`](https://preactjs.com/guide/v10/differences-to-react#arguments-in-componentrender). Preact always has two agruments. Pros: works in both 8.x and 10.x, it's an official documented difference and it is less likely to be changed in the current release Cons: it is not completely synchronous, however it doesn't output an html element.

Let me know what you think, 
Cheers.

### PR DESCRIPTION
**Pull request by molefrog**
Resolves #59 

This PR implements `<WithEnvironment />` component which can be used to detect the current env. It has an improved check for Preact that relies on the [number of arguments passed down to `render()`](https://preactjs.com/guide/v10/differences-to-react#arguments-in-componentrender) and it doesn't require a re-render. I haven't wired it up with the rest of the code yet (I plan to do that in the next PR), as there are some caveats we need to discuss first:

1. It uses a `class` keyword which could potentially increase the bundle size since we target es5 in the config. I don't think it would create any issues though, but proper bundle analysis is probably required 
2. There is no proper test to check if the argument condition does indeed return `true` when rendered via Preact. I guess that's the responsibility of a more higher-level integration test, ideally covered by `<FpjsProvider />` tests, but I didn't want to overload the PR and wanted to make it more atomic.

### CODE REVIEW

File: src/detect-env.ts [hunk 1]
```diff
@@ -1,7 +1,8 @@
 import { Env, type EnvDetails } from './env.types'
 
 export interface DetectEnvContext {
-  syntheticEventDetected: boolean
+  syntheticEventDetected?: boolean
```
Comment (molefrog) [inline src/detect-env.ts:4]: That would be removed once `<FpjsProvider />` starts using `WithEnvironment` under the hood.

**Comment by TheUnderScorer** [comment]
Thanks for the PR @molefrog! I don't have any comments so far, everything looks good.

Let's keep this PR as a "feature branch" so that you can target your next PRs to this one, and once everything is ready we can merge it into a single release.

**Comment by molefrog** [comment]
> Let's keep this PR as a "feature branch" so that you can target your next PRs to this one, and once everything is ready we can merge it into a single release.

Good, I was actually going to propose the same thing. I'll try to get the next one ready early next week.

**Comment by molefrog** [comment]
Hi @TheUnderScorer, I've just updated the branch with the changes we discussed:
- `FpjsProvider` now uses `WithEnvironment` internally. This allowed me to simplify the initialisation logic and now the client is only initialised on a) the first render b) when the config changes if `forceRebuild` is set to `true`
- I covered the new Preact detection logic with a simple test. Unfortunately, having two test suites in one file for React and Preact wasn't possible, there are some logic involved with the way modules are required. So I had to introduce a file `with-environment.preact.test.tsx` not sure if it matches the naming conventions so I'm open to feedback.

File: src/detect-env.ts [hunk 2]
```diff
@@ -29,18 +29,16 @@ function runEnvChecks(...strategies: EnvCheckStrategy[]) {
...
+  return context.classRenderReceivesAnyArguments
 }
 
 /**
  * Checks if user is using react.
- * So far we are doing that by checking if synthetic event was detected.
  * */
 function isReact(context: DetectEnvContext) {
-  return context.syntheticEventDetected
+  return !isPreact(context)
```
Comment (molefrog) [inline src/detect-env.ts:41]: Not sure if we want to be more explicit here and have something like:
```suggestion
  return !context.classRenderReceivesAnyArguments
```

File: src/detect-env.ts [hunk 3]
```diff
@@ -1,7 +1,7 @@
 import { Env, type EnvDetails } from './env.types'
 
 export interface DetectEnvContext {
-  syntheticEventDetected: boolean
+  classRenderReceivesAnyArguments?: boolean
```
[the remark this answers is not in the document]
Comment (TheUnderScorer) [reply src/detect-env.ts:4]: Sounds good to me!

**Comment by TheUnderScorer** [comment]
Thanks for the update @molefrog!

I like the preact test suite implementation, nice work!

File: src/detect-env.ts (hunk 2 above)
Comment (molefrog) [reply src/detect-env.ts]: Good, done.

**Comment by molefrog** [comment]
@molefrog I've fixed all of the comments above, including the removal of `SyntheticEventDetector` which is not being used anymore in the code.

**Review by ilfa — APPROVED** [review]

**Comment by TheUnderScorer** [comment]
Thanks, @molefrog! From your perspective, do you think that there is anything left to do in regard of this PR?

**Comment by molefrog** [comment]
> Thanks, @molefrog! From your perspective, do you think that there is anything left to do in regard of this PR?

Well, I have tested all examples in the repo with a real API key, works as expected. 

There is one additional commit I haven't added here, which is supposed to add an extra check before we call `cloneElement` in `WithEnvironment`. Calling `cloneElement` on plain children (text, booleans etc.) breaks the render, however our types don't allow such children. 

https://github.com/molefrog/fingerprintjs-pro-react/commit/215d674ad462e6696a8605501d3cd4c638fc50d8

I think we can safely ignore this, but it is up to you.

**Comment by molefrog** [comment]
Hi! Just curious if there is anything else that I should consider working on in this PR 😃 I have some ideas for further improvements, but I just don't want overload this PR.

**Comment by TheUnderScorer** [comment]
Hel @molefrog! 

For now, I think that we can consider it finished 😃. Sorry for the overall delay, I was on holiday for the last couple of days.

**Review by TheUnderScorer — APPROVED** [review]

**Pull request merged**